File: lib/core/di/injection_container.dart
import 'package:get_it/get_it.dart';

import '../../data/database/app_database.dart';
import '../../data/repositories/local_branch_repository.dart';
import '../../data/repositories/local_commit_repository.dart';
import '../../data/repositories/local_repository_repository.dart';
import '../../data/repositories/local_task_repository.dart';
import '../../domain/repositories/i_branch_repository.dart';
import '../../domain/repositories/i_commit_repository.dart';
import '../../domain/repositories/i_repository_repository.dart';
import '../../domain/repositories/i_task_repository.dart';
import '../../domain/services/data_export_service.dart';
import '../../domain/usecases/branch/create_branch_usecase.dart';
import '../../domain/usecases/branch/delete_branch_usecase.dart';
import '../../domain/usecases/branch/merge_branch_usecase.dart';
import '../../domain/usecases/repository/create_repository_usecase.dart';
import '../../domain/usecases/repository/delete_repository_usecase.dart';
import '../../domain/usecases/repository/update_repository_usecase.dart';
import '../../domain/usecases/task/complete_task_usecase.dart';
import '../../domain/usecases/task/create_task_usecase.dart';
import '../../domain/usecases/task/delete_task_usecase.dart';
import '../../domain/usecases/task/update_task_usecase.dart';

/// 全局 GetIt 实例
final getIt = GetIt.instance;

/// 配置依赖注入
/// 所有 Repository 通过接口注册，Presentation 层只依赖接口
void configureDependencies() {
  // ── Database ──────────────────────────────────────────────
  getIt.registerLazySingleton(() => AppDatabase());

  // ── Repositories (接口 → 实现) ───────────────────────────
  getIt.registerLazySingleton<IRepositoryRepository>(
    () => LocalRepositoryRepository(getIt<AppDatabase>()),
  );
  getIt.registerLazySingleton<IBranchRepository>(
    () => LocalBranchRepository(getIt<AppDatabase>()),
  );
  getIt.registerLazySingleton<ITaskRepository>(
    () => LocalTaskRepository(getIt<AppDatabase>()),
  );
  getIt.registerLazySingleton<ICommitRepository>(
    () => LocalCommitRepository(getIt<AppDatabase>()),
  );

  // ── Domain Services ───────────────────────────────────────
  getIt.registerLazySingleton(
    () => DataExportService(
      getIt<IRepositoryRepository>(),
      getIt<IBranchRepository>(),
      getIt<ITaskRepository>(),
      getIt<ICommitRepository>(),
    ),
  );

  // ── Use Cases — Repository ────────────────────────────────
  getIt.registerFactory(
    () => CreateRepositoryUseCase(getIt<IRepositoryRepository>()),
  );
  getIt.registerFactory(
    () => UpdateRepositoryUseCase(getIt<IRepositoryRepository>()),
  );
  getIt.registerFactory(
    () => DeleteRepositoryUseCase(getIt<IRepositoryRepository>()),
  );

  // ── Use Cases — Branch ────────────────────────────────────
  getIt.registerFactory(
    () => CreateBranchUseCase(getIt<IBranchRepository>()),
  );
  getIt.registerFactory(
    () => MergeBranchUseCase(
      getIt<IBranchRepository>(),
      getIt<ITaskRepository>(),
      getIt<ICommitRepository>(),
    ),
  );
  getIt.registerFactory(
    () => DeleteBranchUseCase(getIt<IBranchRepository>()),
  );

  // ── Use Cases — Task ──────────────────────────────────────
  getIt.registerFactory(
    () => CreateTaskUseCase(
      getIt<ITaskRepository>(),
      getIt<ICommitRepository>(),
    ),
  );
  getIt.registerFactory(
    () => UpdateTaskUseCase(
      getIt<ITaskRepository>(),
      getIt<ICommitRepository>(),
    ),
  );
  getIt.registerFactory(
    () => CompleteTaskUseCase(
      getIt<ITaskRepository>(),
      getIt<ICommitRepository>(),
    ),
  );
  getIt.registerFactory(
    () => DeleteTaskUseCase(
      getIt<ITaskRepository>(),
      getIt<ICommitRepository>(),
    ),
  );
}

File: web/src/core/di/injection-container.ts
import 'reflect-metadata';
import { container } from 'tsyringe';

import { AppDatabase, appDatabase } from '../../data/db/app-database';
import { DexieBranchRepository } from '../../data/repositories/dexie-branch-repository';
import { DexieCommitRepository } from '../../data/repositories/dexie-commit-repository';
import { DexieRepositoryRepository } from '../../data/repositories/dexie-repository-repository';
import { DexieTaskRepository } from '../../data/repositories/dexie-task-repository';

import { IBranchRepository } from '../../domain/repositories/i-branch-repository';
import { ICommitRepository } from '../../domain/repositories/i-commit-repository';
import { IRepositoryRepository } from '../../domain/repositories/i-repository-repository';
import { ITaskRepository } from '../../domain/repositories/i-task-repository';
import { DataExportService } from '../../domain/services/data-export-service';

import { CreateBranchUseCase } from '../../application/usecases/branch/create-branch-usecase';
import { DeleteBranchUseCase } from '../../application/usecases/branch/delete-branch-usecase';
import { MergeBranchUseCase } from '../../application/usecases/branch/merge-branch-usecase';
import { CreateRepositoryUseCase } from '../../application/usecases/repository/create-repository-usecase';
import { DeleteRepositoryUseCase } from '../../application/usecases/repository/delete-repository-usecase';
import { UpdateRepositoryUseCase } from '../../application/usecases/repository/update-repository-usecase';
import { CompleteTaskUseCase } from '../../application/usecases/task/complete-task-usecase';
import { CreateTaskUseCase } from '../../application/usecases/task/create-task-usecase';
import { DeleteTaskUseCase } from '../../application/usecases/task/delete-task-usecase';
import { UpdateTaskUseCase } from '../../application/usecases/task/update-task-usecase';
import { ImportDataUseCase } from '../../application/usecases/import-data-usecase';

import { WebFileSaveService } from '../../platform/web-file-save-service';
import { WebNotificationService } from '../../platform/web-notification-service';

container.registerInstance(AppDatabase, appDatabase);

container.register<IRepositoryRepository>('IRepositoryRepository', {
  useFactory: () => new DexieRepositoryRepository(container.resolve(AppDatabase)),
});

container.register<IBranchRepository>('IBranchRepository', {
  useFactory: () => new DexieBranchRepository(container.resolve(AppDatabase)),
});

container.register<ITaskRepository>('ITaskRepository', {
  useFactory: () => new DexieTaskRepository(container.resolve(AppDatabase)),
});

container.register<ICommitRepository>('ICommitRepository', {
  useFactory: () => new DexieCommitRepository(container.resolve(AppDatabase)),
});

container.register(DataExportService, {
  useFactory: () =>
    new DataExportService(
      container.resolve<IRepositoryRepository>('IRepositoryRepository'),
      container.resolve<IBranchRepository>('IBranchRepository'),
      container.resolve<ITaskRepository>('ITaskRepository'),
      container.resolve<ICommitRepository>('ICommitRepository')
    ),
});

container.register(CreateRepositoryUseCase, {
  useFactory: () =>
    new CreateRepositoryUseCase(
      container.resolve<IRepositoryRepository>('IRepositoryRepository'),
      container.resolve<IBranchRepository>('IBranchRepository')
    ),
});

container.register(UpdateRepositoryUseCase, {
  useFactory: () =>
    new UpdateRepositoryUseCase(
      container.resolve<IRepositoryRepository>('IRepositoryRepository')
    ),
});

container.register(DeleteRepositoryUseCase, {
  useFactory: () =>
    new DeleteRepositoryUseCase(
      container.resolve<IRepositoryRepository>('IRepositoryRepository'),
      container.resolve<IBranchRepository>('IBranchRepository'),
      container.resolve<ITaskRepository>('ITaskRepository')
    ),
});

container.register(CreateBranchUseCase, {
  useFactory: () =>
    new CreateBranchUseCase(
      container.resolve<IBranchRepository>('IBranchRepository')
    ),
});

container.register(MergeBranchUseCase, {
  useFactory: () =>
    new MergeBranchUseCase(
      container.resolve<IBranchRepository>('IBranchRepository'),
      container.resolve<ITaskRepository>('ITaskRepository'),
      container.resolve<ICommitRepository>('ICommitRepository')
    ),
});

container.register(DeleteBranchUseCase, {
  useFactory: () =>
    new DeleteBranchUseCase(
      container.resolve<IBranchRepository>('IBranchRepository')
    ),
});

container.register(CreateTaskUseCase, {
  useFactory: () =>
    new CreateTaskUseCase(
      container.resolve<ITaskRepository>('ITaskRepository'),
      container.resolve<ICommitRepository>('ICommitRepository')
    ),
});

container.register(UpdateTaskUseCase, {
  useFactory: () =>
    new UpdateTaskUseCase(
      container.resolve<ITaskRepository>('ITaskRepository'),
      container.resolve<ICommitRepository>('ICommitRepository')
    ),
});

container.register(CompleteTaskUseCase, {
  useFactory: () =>
    new CompleteTaskUseCase(
      container.resolve<ITaskRepository>('ITaskRepository'),
      container.resolve<ICommitRepository>('ICommitRepository')
    ),
});

container.register(DeleteTaskUseCase, {
  useFactory: () =>
    new DeleteTaskUseCase(
      container.resolve<ITaskRepository>('ITaskRepository'),
      container.resolve<ICommitRepository>('ICommitRepository')
    ),
});

container.register(ImportDataUseCase, {
  useFactory: () =>
    new ImportDataUseCase(
      container.resolve<IRepositoryRepository>('IRepositoryRepository'),
      container.resolve<IBranchRepository>('IBranchRepository'),
      container.resolve<ITaskRepository>('ITaskRepository'),
      container.resolve<ICommitRepository>('ICommitRepository')
    ),
});

container.registerInstance(WebFileSaveService, new WebFileSaveService());
container.registerInstance(WebNotificationService, new WebNotificationService());

export { container };
